### PR TITLE
Fix aesmd service permission in linksgx.sh.

### PR DESCRIPTION
**Pull request by yuguorui**
Currently sgx-aesm-service does not depend on any other sgx packages, which makes it difficult to guarantee the installation order of sgx-aesm-service. Specifically, sgx-aesm-service may be installed earlier than libsgx-ae-pce, but the user group `sgx_prv` was created by the latter.

This PR ensures that no matter what the order of package installation is, aesmd can have access to the sgx_provision device file.

### CODE REVIEW

**Comment by fqiu1** [comment]
It's legitimate to install sgx-aesm-service without creating sgx_prv group. For instance, some users might just want to load normal enclaves and don't want to use attestation.

**Comment by yuguorui** [comment]
> It's legitimate to install sgx-aesm-service without creating sgx_prv group. For instance, some users might just want to load normal enclaves and don't want to use attestation.

Yes, you are right in non-DCAP platform.
But for DCAP scenarios, which runs upon CPUs support FLC, I do not find a situation where aesmd-service needs to be installed without attestation.

File: linux/installer/common/sgx-aesm-service/linksgx.sh [hunk 1]
```diff
@@ -35,8 +35,20 @@ if test $(id -u) -ne 0; then
     exit 1
 fi
 
-/usr/bin/getent group sgx_prv &> /dev/null
-if [ $? == "0" ]; then
+if [ -c /dev/sgx_provision -o -c /dev/sgx/provision ]; then
+    /usr/bin/getent group sgx_prv &> /dev/null
```
Comment (veehaitch) [inline linux/installer/common/sgx-aesm-service/linksgx.sh:39]:
```suggestion
    getent group sgx_prv &> /dev/null
```

File: linux/installer/common/sgx-aesm-service/linksgx.sh [hunk 2]
```diff
@@ -35,8 +35,20 @@ if test $(id -u) -ne 0; then
...
+    if [ "$?" != "0" ]; then
+        # Add sgx_prv for dcap driver, which ensures that no matter what 
+        # the order of package installation, aesmd can have access to 
+        # the sgx_provision device file.
+        /usr/sbin/groupadd sgx_prv
```
Comment (veehaitch) [inline linux/installer/common/sgx-aesm-service/linksgx.sh:44]:
```suggestion
        groupadd sgx_prv
```

File: linux/installer/common/sgx-aesm-service/linksgx.sh [hunk 3]
```diff
@@ -35,8 +35,20 @@ if test $(id -u) -ne 0; then
...
+
+        if ! which udevadm &> /dev/null; then
+            return 0
+        fi
+        udevadm control --reload || :
+        udevadm trigger || :
+    fi
     /usr/sbin/usermod -aG sgx_prv aesmd &> /dev/null
```
Comment (veehaitch) [inline linux/installer/common/sgx-aesm-service/linksgx.sh:52]:
```suggestion
    usermod -aG sgx_prv aesmd &> /dev/null
```

**Review by veehaitch — CHANGES_REQUESTED** [review]
Please avoid absolute paths to executables. This assumes the host system adheres to the Filesystem Hierarchy Standard which is not always the case (a notable exception is Nix). Also see #730.

**Comment by yuguorui** [comment]
> Please avoid absolute paths to executables. This assumes the host system adheres to the Filesystem Hierarchy Standard which is not always the case (a notable exception is Nix). Also see #730.

Sure, thanks for your comments. 

I will push a new version later.

File: linux/installer/common/sgx-aesm-service/linksgx.sh [hunk 4]
```diff
@@ -35,8 +35,20 @@ if test $(id -u) -ne 0; then
...
+if [ -c /dev/sgx_provision -o -c /dev/sgx/provision ]; then
+    /usr/bin/getent group sgx_prv &> /dev/null
+    if [ "$?" != "0" ]; then
+        # Add sgx_prv for dcap driver, which ensures that no matter what 
+        # the order of package installation, aesmd can have access to 
+        # the sgx_provision device file.
+        /usr/sbin/groupadd sgx_prv
+
+        if ! which udevadm &> /dev/null; then
+            return 0
```
Comment (lzha101) [inline linux/installer/common/sgx-aesm-service/linksgx.sh:47]: Better to use `exit 0` instead of `return 0`.

Comment (yuguorui) [reply linux/installer/common/sgx-aesm-service/linksgx.sh:47]: Copy that.

Comment (yuguorui) [reply linux/installer/common/sgx-aesm-service/linksgx.sh]: Fixed.

**Pull request merged**